runtime/jit_vm/cthelpers.cpp

```diff
@@ -78,6 +78,48 @@ jitGetInterfaceITableIndexFromCP(J9VMThread *currentThread, J9ConstantPool *cons
 	return interfaceClass;
 }
 
+/* Only returns non-null if the method is not to be dispatched by itable, i.e.
+ * if it is:
+ * - private, using direct dispatch;
+ * - a final method of Object, using direct dispatch; or
+ * - a non-final method of Object, using virtual dispatch.
+ */
+J9Method*
+jitGetImproperInterfaceMethodFromCP(J9VMThread *currentThread, J9ConstantPool *constantPool, UDATA cpIndex)
+{
+	J9RAMInterfaceMethodRef *ramMethodRef = (J9RAMInterfaceMethodRef*)constantPool + cpIndex;
+	/* interfaceClass is used to indicate 'resolved'. Make sure to read it first */
+	J9Class* volatile interfaceClass = (J9Class*)ramMethodRef->interfaceClass;
+	VM_AtomicSupport::readBarrier();
+	UDATA volatile methodIndexAndArgCount = ramMethodRef->methodIndexAndArgCount;
+	J9Method *improperMethod = NULL;
+	/* If the ref is resolved, do not call the resolve helper, as it will currently fail if the interface class is not initialized */
+	if (NULL == interfaceClass) {
+		J9RAMInterfaceMethodRef localEntry;
+		if (NULL == currentThread->javaVM->internalVMFunctions->resolveInterfaceMethodRefInto(currentThread, constantPool, cpIndex, J9_RESOLVE_FLAG_JIT_COMPILE_TIME, &localEntry)) {
+			goto done;
+		}
+		interfaceClass = (J9Class*)localEntry.interfaceClass;
+		methodIndexAndArgCount = localEntry.methodIndexAndArgCount;
+	}
+	if (J9_ARE_ANY_BITS_SET(methodIndexAndArgCount, J9_ITABLE_INDEX_TAG_BITS)) {
+		UDATA methodIndex = (methodIndexAndArgCount & ~J9_ITABLE_INDEX_TAG_BITS) >> 8;
+		J9Class *jlObject = J9VMJAVALANGOBJECT_OR_NULL(currentThread->javaVM);
+		if (J9_ARE_NO_BITS_SET(methodIndexAndArgCount, J9_ITABLE_INDEX_METHOD_INDEX)) {
+			/* Object method in the vTable */
+			improperMethod = *(J9Method**)((UDATA)jlObject + methodIndex);
+		} else if (J9_ARE_ANY_BITS_SET(methodIndexAndArgCount, J9_ITABLE_INDEX_OBJECT)) {
+			/* Object method not in the vTable */
+			improperMethod = jlObject->ramMethods + methodIndex;
+		} else {
+			/* Private interface method */
+			improperMethod = interfaceClass->ramMethods + methodIndex;
+		}
+	}
+done:
+	return improperMethod;
```

Comment:
How does this handle improper but not yet resolved methods?  The current implementation returns null which would imply, given the function's contract, that this is a proper method but it might not be when it's resolved.

Reply:
Unresolved is an entirely different path. This code applies only to the resolved case.

Reply:
This code is actually incorrect - there's no way to determine the method in the virtual case without knowing the receiver class. For reference, the "improper" case from the interpreter is:

```
						/* Object or private interface method invoke */
						if (J9_ARE_ANY_BITS_SET(methodIndexAndArgCount, J9_ITABLE_INDEX_METHOD_INDEX)) {
							if (J9_ARE_ANY_BITS_SET(methodIndexAndArgCount, J9_ITABLE_INDEX_OBJECT)) {
								/* Object method not in the vTable */
								_sendMethod = J9VMJAVALANGOBJECT_OR_NULL(_vm)->ramMethods + methodIndex;
							} else {
								/* Private interface method */
								_sendMethod = interfaceClass->ramMethods + methodIndex;
							}
						} else {
							/* Object method in the vTable */
							_sendMethod = *(J9Method**)((UDATA)receiverClass + methodIndex);
						}```

Reply:
In the virtual case, the compiler uses the method to get a VFT offset, and generates a virtual call

Reply:
OK, I understand now.

Reply:
I'll approve - I know Dan would like the code reordered to match the interpreter cases.

Reply:
OK, reordering to match the interpreter

Reply:
@jdmpapin can you add a new commit? We can avoid re-running the testing if we can manually inspect that the change was just simple code movement.

Reply:
Can you also update the comments to include this info in the code?  It'll save someone a lot of digging and frustration in the future if we can record this now while it's fresh.

Reply:
Updated